### ISSUE TITLE
Consider merging both SDKs into one SDK, or into one Xcode project

### ISSUE THREAD

**Issue by ben-xD** [ably/ably-asset-tracking-swift#154]
There are 2 separate questions here: 
- Why do we have 2 SDKs: **Publisher**, and **Subscriber**? These SDKs share code using the `Core` Xcode project. The Ably-cocoa SDK is a single-project repository. Could we merge SDKs into one?
- If we do still want 2 SDKs, why are they in separate Xcode **projects**? We could put them in the same project, but different **targets**.

## Current problems
- Manual installation of the SDKs will be harder and quite involved for customers. My brief evaluation of this is: SDK users need to add 3 projects to their workspace, set up cocoapods dependencies (modify their cocoapods files, technical a 4th xcode project), and **manually** add each file they use from the SDK to `Compile Sources….` in `Build Phases`.
- Development is tedious: new files need to be added to each individual project (Subscriber, Publisher) in the Xcode workspace. This will catch developers who don't know about this (it took me a good few hours to find that I needed to add the specific file to every project I want to compile): e.g. I faced it, and this was also the reason Quintin's PR did not build in: https://github.com/ably/ably-asset-tracking-cocoa/pull/143
  - There is a previous issue raised for complexity in Xcode projects before: https://github.com/ably/ably-asset-tracking-cocoa/issues/43
  - I'm not sure how this will be translated when supporting Swift Package Manager.

## Code breakdown
I counted lines of code using `find -E . -iregex ".*\.(dart|m|java|kt|swift)" | xargs wc`
- Core SDK: 1769 lines
- Publisher SDK: 3529 lines
- Subscriber SDK: 1276 lines

## Benefits: 
- Solves the problems listed above (installation by users and development experience)
- Reduced complexity/ work for SDK users: Users just decide they want asset tracking, and they add the ably-asset-tracking-swift library to their app. They don't have to choose between them, or add a second one in the future (if they want both functionality)
- ~~We don't need to release multiple SDKs to Cocoapods (or configure more settings in `Package.swift` for SPM support).~~
- ~~We don't need to coordinate changes to the `Core` (shared code) Xcode project~~

## Disadvantages: 
- @QuintinWillison, do you have any ideas on more disadvantages, I struggled to come up with these (which may/ may not not be significant):
  - Decoupled release: we can release features separately to the Publisher and Subscriber SDK (I don't see why we don't want to release them together though)
  - Reduced library size: Customer apps which only Subscribe can use the Subscribe SDK (This optimisation is not needed? These extra lines of code do not translate to significant kilobytes of increased app size.)

## TODOs:
- Discuss if we can merge the SDKs into one SDK.
- Discuss if we can move all SDK projects into 1 project (separate Xcode project for examples). Potentially have different targets, or just have 1 target?

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/SDK-900) by [Unito](https://www.unito.io)

**Comment by QuintinWillison** [comment]
@robertkuraj / @mmiedlarz - can either of you add any context or thoughts that might help, please?
FYI, @KacperKluka

**Comment by mmiedlarz** [comment]
Hey @QuintinWillison, sure thing. 
As far as I remember, it was a hard initial SDK requirement to be installed as separate CocoaPod pods (so Publisher users don't need to import Subscriber and vice versa), and for the initial launch, we considered CocoaPods as only way to install SDK. 
If those requirements are no longer valid, I totally agree that merging Publisher and Subscriber will make lot of things easier - i.e. we could remove [this](https://github.com/ably/ably-asset-tracking-cocoa#working-on-code-shared-between-publisher-and-subscriber) step from the readme.

**Comment by paddybyers** [comment]
The subscriber SDK has to be usable without any mapbox dependency

**Comment by ben-xD** [comment]
> The subscriber SDK has to be usable without any mapbox dependency

Okay, that answers the separate SDK problem then, I think it makes sense to ship 2 SDKs. However, **I believe** we can still reduce the complexity (and solve some of the problems listed) by having 1 Xcode project instead of 4 projects in 1 workspace (`Core`, `Publisher`, `Subscriber`, `Pods`). The only target that will depend on Mapbox Navigation will be `Publisher`.

Having said that, there is a related issue that was closed: **Use .xcodeproj instead of targets for managing SDKs**  https://github.com/ably/ably-asset-tracking-cocoa/issues/34 @mmiedlarz , could you give more information about this ticket, there is not much info on it in Jira or Github:

**Comment by ben-xD** [comment]
Small update:

**TLDR:** This issue is not urgent, but it _might_ block SPM support or make it difficult to achieve (unclear). Lets look at is as part of supporting SPM ticket: https://github.com/ably/ably-asset-tracking-cocoa/issues/148

- I've added Cocoapods user installation instructions (via GitHub, not Cocoapods.org to avoid releasing it), so the existing structure of multiple Xcode projects in 1 workspace allowed this to work (though it took me quite a few hours to get working): PR is https://github.com/ably/ably-asset-tracking-cocoa/issues/155
- If we support only Swift Package Manager, we can simplify the setup into 1 Xcode project + multiple targets. Most Xcode repos are single projects, and this gives the benefits listed previously. Repos which use Cocoapods are are multi-project workspaces. SPM is the recommended way to release software for Cocoa, but I'm not sure how the SPM will work for multiple Xcode projects.

**Comment by lukasz-szyszkowski** [comment]
I think this issue is not valid anymore. I recommend closing it @jamienewcomb @AndyNicks

**Comment by QuintinWillison** [comment]
Agreed. Thanks, @lukasz-szyszkowski
